include/itkMaxPhaseCorrelationOptimizer.hxx

```diff
@@ -82,7 +77,125 @@ MaxPhaseCorrelationOptimizer< TRegistrationMethod >
     return;
     }
 
-  m_MaxCalculator->SetImage( input );
+  const typename ImageType::RegionType lpr = input->GetLargestPossibleRegion();
+  const typename ImageType::SizeType size = lpr.GetSize();
+  const typename ImageType::IndexType oIndex = lpr.GetIndex();
+
+  const typename ImageType::SpacingType spacing = input->GetSpacing();
+  const typename ImageType::PointType fixedOrigin = fixed->GetOrigin();
+  const typename ImageType::PointType movingOrigin = moving->GetOrigin();
+
+  // create the image which be biased towards the expected solution and be zero-suppressed
+  typename ImageType::Pointer iAdjusted = ImageType::New();
+  iAdjusted->CopyInformation( input );
+  iAdjusted->SetRegions( input->GetBufferedRegion() );
+  iAdjusted->Allocate( false );
+
+  typename ImageType::IndexType adjustedSize;
+  typename ImageType::IndexType directExpectedIndex;
+  typename ImageType::IndexType mirrorExpectedIndex;
+  double distancePenaltyFactor = 0.0;
+  for ( unsigned d = 0; d < ImageDimension; d++ )
+    {
+    adjustedSize[d] = size[d] + oIndex[d];
+    distancePenaltyFactor += adjustedSize[d] * adjustedSize[d]; // make it proportional to image size
+    directExpectedIndex[d] = ( movingOrigin[d] - fixedOrigin[d] ) / spacing[d] + oIndex[d];
+    mirrorExpectedIndex[d] = ( movingOrigin[d] - fixedOrigin[d] ) / spacing[d] + adjustedSize[d];
+    }
+
+  distancePenaltyFactor = -m_BiasTowardsExpected / distancePenaltyFactor;
+
+  MultiThreaderBase* mt = this->GetMultiThreader();
+  mt->ParallelizeImageRegion<ImageDimension>( lpr,
+    [&](const typename ImageType::RegionType & region)
+    {
+      ImageRegionConstIterator< ImageType > iIt(input, region);
+      ImageRegionIteratorWithIndex< ImageType > oIt(iAdjusted, region);
+      for (; !oIt.IsAtEnd(); ++iIt, ++oIt)
+        {
+        typename ImageType::IndexType ind = oIt.GetIndex();
+        IndexValueType dist = 0;
+        for (unsigned d = 0; d < ImageDimension; d++)
+          {
+          IndexValueType distDirect = ( directExpectedIndex[d] - ind[d] ) * ( directExpectedIndex[d] - ind[d] );
+          IndexValueType distMirror = ( mirrorExpectedIndex[d] - ind[d] ) * ( mirrorExpectedIndex[d] - ind[d] );
+          if ( distDirect <= distMirror )
+            {
+            dist += distDirect;
+            }
+          else
+            {
+            dist += distMirror;
+            }
+          }
+
+        typename ImageType::PixelType pixel = 1000*iIt.Get() * std::exp( distancePenaltyFactor * dist ); //TODO: remove 1000 factor
+        oIt.Set( pixel );
+        }
+    },
+    nullptr );
+
+  WriteDebug( iAdjusted.GetPointer(), "iAdjusted.nrrd" );
```

Comment:
Debug code

Reply:
We can't remove this debugging just yet. We still need it.